pkg/movingaverage/median_filter.go

```diff
@@ -14,39 +14,185 @@
 
 package movingaverage
 
-import "github.com/elliotchance/pie/v2"
+import (
+	"errors"
+	"math"
+)
 
 // MedianFilter works as a median filter with specified window size.
 // There are at most `size` data points for calculating.
 // References: https://en.wikipedia.org/wiki/Median_filter.
+// Note: MedianFilter is not Thread-Safety.
+// If the some items are `math.MaxFloat64`, no guarantee that the results are correct.
 type MedianFilter struct {
-	// It is not thread safe to read and write records at the same time.
-	// If there are concurrent read and write, the read may get an old value.
-	// And we should avoid concurrent write.
 	records []float64
 	size    uint64
 	count   uint64
-	result  float64
+	// g is used to count the number of values which are greater than median
+	g uint64
+	// l is used to count the number of values which are less than median
+	l      uint64
+	result float64
 }
 
 // NewMedianFilter returns a MedianFilter.
 func NewMedianFilter(size int) *MedianFilter {
 	return &MedianFilter{
 		records: make([]float64, size),
 		size:    uint64(size),
-		result:  0,
 	}
 }
 
+// err is used to avoid returning MaxFloat64 for concurrency
+func (r *MedianFilter) findTwoMinNumber(needSecond bool) (first, second float64, err error) {
+	len := r.size
+	if r.count < r.size {
+		len = r.count
+	}
+	first = math.MaxFloat64
+	second = math.MaxFloat64
+	var pos uint64
+	for i := uint64(0); i < len; i++ {
+		if r.records[i] > r.result && r.records[i] < first {
+			first = r.records[i]
+			pos = i
+		}
+	}
+	if needSecond {
```

Comment:
how about split it to two function `findSecondMinNumber` and `findMinNumber`  without bool control